packages/cinterop/src/nativeDarwin/kotlin/io/realm/kotlin/internal/interop/RealmInterop.kt

```diff
@@ -1005,24 +980,26 @@ actual object RealmInterop {
         }
     }
 
-    actual fun realm_set_get(set: RealmSetPointer, index: Long): RealmValue {
-        memScoped {
-            val cvalue = alloc<realm_value_t>()
-            checkedBooleanResult(
-                realm_wrapper.realm_set_get(set.cptr(), index.toULong(), cvalue.ptr)
-            )
-            return from_realm_value(cvalue)
-        }
+    // Returning a non-nullable transport here goes against the approach that increases
+    // performance (since we need to call getType on the transport object). This is needed though
+    // because this function is called when calling 'iterator.remove' and causes issues when telling
+    // the C-API to delete a null transport created within the scope. We need to investigate further
+    // how to improve this.
+    actual fun MemAllocator.realm_set_get(set: RealmSetPointer, index: Long): RealmValue {
+        val struct = allocRealmValueT()
+        checkedBooleanResult(realm_wrapper.realm_set_get(set.cptr(), index.toULong(), struct.ptr))
+        return RealmValue(struct)
     }
 
-    actual fun realm_set_find(set: RealmSetPointer, value: RealmValue): Boolean {
+    actual fun realm_set_find(set: RealmSetPointer, transport: RealmValue): Boolean {
+        // TODO optimize: reuse the same memory allocation
```

Comment:
This is something that would vote for Clementes approach of keeping the `value: RealmValue` allocation/assignment inside the cinterop layer as a lambda block which would tie the `value` memory scope inside this internal scope. But it could also be that we could just use context receivers to ensure that this happens inside an external `memScope` if that is possible to differentiate on each platform.